src/expr/impl/src/table_function/unnest.rs

```diff
@@ -21,5 +21,5 @@ use risingwave_expr::function;
     type_infer = "|args| Ok(args[0].unnest_list().clone())"
 )]
 fn unnest(list: ListRef<'_>) -> impl Iterator<Item = Option<ScalarRefImpl<'_>>> {
-    list.flatten().into_iter()
+    list.flatten().iter()
```

Comment:
Is this still suboptimal? We are still iterating over the list slices. Can we directly reinterpret the whole `ListArray` to the underlying `ArrayImpl` without iteration? 🤔

(BTW, I come here when I was thinking about expressions for the `map` type and found we don't seem to have vectorized exprs for `list` also. sth like https://github.com/apache/datafusion/pull/7825/files#diff-1cddf6dfddea9d7c2e299e973bf915176256222281028f2571b1662ed92cd58dR190-R195

Reply:
> We are still iterating over the list slices. Can we directly reinterpret the whole `ListArray` to the underlying `ArrayImpl` without iteration? 🤔

I don't quite get it. Can you elaborate more?

Reply:
Did you see `map_inner`?

Reply:
`map_inner` manipulates the underlying `entires` child array of the list array directly. Generally, I feel we haven't utilize such optimization a lot. Not specific to `unnest`.

Reply:
> Not specific to `unnest`.

I wasn't specifically talking about `unnest`. I'm referring to the vectorized execution of any set-returning functions for `list` type. Not manipulating the `offsets` array only works if we don't need to know the output length of each `list` scalar, which requires that there's no other set-returning or scalar functions in the same `ProjectSet`.

Reply:
Hmmm, actually I was only thinking about scalar functions.

But I'm not sure I understand what you mean. I feel it also work for `ProjectSet`:

Specifically, `TableFunction` will return a `chunk` with 2 columns (`row_idx`, and `values` ).

https://github.com/risingwavelabs/risingwave/blob/dcb565eda57ba8742b1e7d4856902335a98ce84f/src/expr/core/src/table_function/mod.rs#L82-L95


For `unnest`:
```
    /// # input chunk     output chunks
    /// [1,2]   --------------> 0 1
    /// [3,4,5] -----┐          0 2
    ///              │    
    ///              └--------> 1 3
    ///                         1 4
    ///                         1 5
    ///                row idx--^ ^--values
```

To build `row_idx`, we still need to iterate over `list_array.iter()` though, but `values` seems to be exactly the same as `list_array.inner`, so we can directly `clone` it. In the current implementation, we have an extra copy when building `values`.

Reply:
> Is this still suboptimal? We are still iterating over the list slices. Can we directly reinterpret the whole `ListArray` to the underlying `ArrayImpl` without iteration? 🤔

Yes, of course. However, this specialization has to be done with `#[build_function]`.

Reply:
> but `values` seems to be exactly the same as `list_array.inner`, so we can directly `clone` it. In the current implementation, we have an extra copy when building `values`.

You're right. There's indeed an extra allocation. But anyhow we still have to do copying when reconstructing the output for a `ProjectSet`.

https://github.com/risingwavelabs/risingwave/blob/0b0cb85f487a77e69dd5bfdd8854b7f4dd11591f/src/stream/src/executor/project_set.rs#L186

On the other hand, if we can have an owned version of `ListRef` (like holding an `ArrayRef` instead of `&'a ArrayImpl`) and directly let `eval` returning an iterator of `ListRef`, we can also achieve zero-copy.